src/helpers/createArchiveLoader.ts

```diff
@@ -1,15 +1,28 @@
-import { memoize } from 'lodash/fp';
-import { ArchiveContent } from '../app/content/types';
+import memoize from 'lodash/fp/memoize';
+import { ArchiveBook, ArchiveContent, ArchivePage } from '../app/content/types';
 
-export default (url: string) => memoize((id: string) => fetch(url + id)
-  .then((response) => {
-    if (response.status !== 200) {
-      return response.text().then((message: string) => {
-        throw new Error(`Error response from archive ${response.status}: ${message}`);
-      });
-    }
-    return response;
-  })
-  .then((response) => response.json())
-  .then((response) => response as ArchiveContent)
-);
+export default (url: string) => {
+  const cache = new Map();
+  const loader = memoize((id: string) => fetch(url + id)
+    .then((response) => {
+      if (response.status !== 200) {
+        return response.text().then((message: string) => {
+          throw new Error(`Error response from archive ${response.status}: ${message}`);
+        });
+      }
+      return response;
+    })
+    .then((response) => response.json() as Promise<ArchiveContent>)
+    .then((response) => {
+      cache.set(id, response);
+      return response;
+    })
+  );
+
+  return {
+    book: (bookId: string) => loader(bookId) as Promise<ArchiveBook>,
```

Comment:
Could this be `=> cache.get(bookId) || loader(bookId)` ? and therefore avoid the need for an explicit `cachedBook`?

Reply:
no the loader returns promises and the cache returns json